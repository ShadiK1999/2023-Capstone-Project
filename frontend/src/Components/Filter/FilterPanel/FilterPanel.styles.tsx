import { style } from 'typestyle';

export const panel = style({
  position: 'absolute',
  display: 'flex',
  flexDirection: 'column',
  gap: '30px',
  top: '110%',
  left: '0',
  width: '150%',
  padding: '15px',
  border: '1px solid black',
  backgroundColor: 'white',
  borderRadius: '5px',
  zIndex: 1,
});
