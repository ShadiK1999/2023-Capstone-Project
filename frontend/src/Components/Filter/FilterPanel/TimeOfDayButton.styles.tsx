import { style } from 'typestyle';

export const container = style({
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'space-evenly',
  alignItems: 'center',
  width: '100%',
});
