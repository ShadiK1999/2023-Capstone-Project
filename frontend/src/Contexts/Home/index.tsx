import { FC, ReactNode, useReducer } from 'react';
import { HomeContext, HomeDispatchContext, initialState } from './state';
import { homeReducer } from './reducer';

type ComponentProps = {
  children: ReactNode;
};

const HomeProvider: FC<ComponentProps> = ({ children }) => {
  const [state, dispatch] = useReducer(homeReducer, initialState);

  return (
    <HomeContext.Provider value={state}>
      <HomeDispatchContext.Provider value={dispatch}>{children}</HomeDispatchContext.Provider>
    </HomeContext.Provider>
  );
};

export { HomeProvider };
