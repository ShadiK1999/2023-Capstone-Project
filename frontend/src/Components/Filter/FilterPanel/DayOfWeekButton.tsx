import { FC, useContext } from 'react';
import { HomeContext, HomeDispatchContext, DayOfWeek } from '../../../Contexts/Home/state';
import { LabelBox } from '../LabelBox';
import { SegmentButton } from '../SegmentButton';
import { produce } from 'immer';

const DayOfWeekButton: FC = () => {
  const dispatch = useContext(HomeDispatchContext);
  const { filter } = useContext(HomeContext);
  const daysOfWeekDisplay: DayOfWeek[] = ['SUN', 'MON', 'TUE', 'WED', 'THU', 'FRI', 'SAT'];

  const handleUpdateDaysOfWeek = (inputDaysOfWeek: string[]) => {
    const updatedFilter = produce(filter, (draft: typeof filter) => {
      draft.daysOfWeek = inputDaysOfWeek as DayOfWeek[];
    });
    dispatch({
      type: 'FILTER_INPUT_CHANGED',
      payload: updatedFilter,
    });
  };

  return (
    <LabelBox label="Opening Days">
      <SegmentButton
        buttons={daysOfWeekDisplay}
        onSelect={handleUpdateDaysOfWeek}
        multiSelect={true}
        initialSelected={filter.daysOfWeek}
      />
    </LabelBox>
  );
};

export { DayOfWeekButton };
