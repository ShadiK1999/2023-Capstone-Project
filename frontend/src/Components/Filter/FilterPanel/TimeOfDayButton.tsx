import { FC, useContext } from 'react';
import * as style from './TimeOfDayButton.styles';
import { HomeContext, HomeDispatchContext } from '../../../Contexts/Home/state';
import { LabelBox } from '../LabelBox';
import { PlusMinusButton } from '../PlusMinusButton';
import { produce } from 'immer';

const TimeOfDayButton: FC = () => {
  const dispatch = useContext(HomeDispatchContext);
  const { filter } = useContext(HomeContext);
  const hoursDisplay: string[] = [
    '12am',
    '1am',
    '2am',
    '3am',
    '4am',
    '5am',
    '6am',
    '7am',
    '8am',
    '9am',
    '10am',
    '11am',
    '12pm',
    '1pm',
    '2pm',
    '3pm',
    '4pm',
    '5pm',
    '6pm',
    '7pm',
    '8pm',
    '9pm',
    '10pm',
    '11pm',
  ];
  const MIN_HOUR = 0;
  const MAX_HOUR = 23;

  const handleUpdateOpenHours = (inputOpenTime: number) => {
    if (inputOpenTime >= filter.closeTime || inputOpenTime < MIN_HOUR) {
      return;
    }

    const updatedFilter = produce(filter, (draft: typeof filter) => {
      draft.openTime = inputOpenTime;
    });

    dispatch({
      type: 'FILTER_INPUT_CHANGED',
      payload: updatedFilter,
    });
  };

  const handleUpdateCloseHours = (inputCloseTime: number) => {
    if (inputCloseTime <= filter.openTime || inputCloseTime > MAX_HOUR) {
      return;
    }

    const updatedFilter = produce(filter, (draft: typeof filter) => {
      draft.closeTime = inputCloseTime;
    });

    dispatch({
      type: 'FILTER_INPUT_CHANGED',
      payload: updatedFilter,
    });
  };

  return (
    <div className={style.container}>
      <LabelBox label="Open from...">
        <PlusMinusButton
          leftButton={{
            onClick: () => handleUpdateOpenHours(filter.openTime - 1),
          }}
          rightButton={{
            onClick: () => handleUpdateOpenHours(filter.openTime + 1),
          }}
        >
          {hoursDisplay[filter.openTime]}
        </PlusMinusButton>
      </LabelBox>
      <LabelBox label="Until...">
        <PlusMinusButton
          leftButton={{
            onClick: () => handleUpdateCloseHours(filter.closeTime - 1),
          }}
          rightButton={{
            onClick: () => handleUpdateCloseHours(filter.closeTime + 1),
          }}
        >
          {hoursDisplay[filter.closeTime]}
        </PlusMinusButton>
      </LabelBox>
    </div>
  );
};

export { TimeOfDayButton };
