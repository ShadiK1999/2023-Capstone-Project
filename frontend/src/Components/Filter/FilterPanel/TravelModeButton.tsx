import { FC, useContext } from 'react';
import { HomeContext, HomeDispatchContext, Filter } from '../../../Contexts/Home/state';
import { LabelBox } from '../LabelBox';
import { produce } from 'immer';
import { SegmentButton } from '../SegmentButton';

const TravelModeButton: FC = () => {
  const travelModes: Filter['mode'][] = ['WALKING', 'CYCLING', 'DRIVING'];
  const dispatch = useContext(HomeDispatchContext);
  const { filter } = useContext(HomeContext);

  const handleUpdateTravelMode = (options: string[]) => {
    const updatedFilter = produce(filter, (draft: typeof filter) => {
      draft.mode = options[0] as Filter['mode'];
    });

    dispatch({
      type: 'FILTER_INPUT_CHANGED',
      payload: updatedFilter,
    });
  };

  return (
    <LabelBox label="Travel Mode">
      <SegmentButton
        buttons={travelModes}
        onSelect={handleUpdateTravelMode}
        multiSelect={false}
        initialSelected={[filter.mode]}
      />
    </LabelBox>
  );
};

export { TravelModeButton };
