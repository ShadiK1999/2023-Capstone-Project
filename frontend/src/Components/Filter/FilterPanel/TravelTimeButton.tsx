import { FC, useContext } from 'react';
import { HomeContext, HomeDispatchContext } from '../../../Contexts/Home/state';
import { PlusMinusButton } from '../PlusMinusButton';
import { produce } from 'immer';
import { LabelBox } from '../LabelBox';

const TravelTimeButton: FC = () => {
  const dispatch = useContext(HomeDispatchContext);
  const { filter } = useContext(HomeContext);

  const handleUpdateTravelTime = (modifier: number) => {
    if (filter.value + modifier >= 45 || filter.value + modifier <= 0) {
      return;
    }

    const updatedFilter = produce(filter, (draft: typeof filter) => {
      draft.value = filter.value + modifier;
    });

    dispatch({
      type: 'FILTER_INPUT_CHANGED',
      payload: updatedFilter,
    });
  };

  return (
    <LabelBox label="Travel Time">
      <PlusMinusButton
        leftButton={{ onClick: () => handleUpdateTravelTime(-5) }}
        rightButton={{ onClick: () => handleUpdateTravelTime(5) }}
      >
        {filter.value + ' min'}
      </PlusMinusButton>
    </LabelBox>
  );
};

export { TravelTimeButton };
