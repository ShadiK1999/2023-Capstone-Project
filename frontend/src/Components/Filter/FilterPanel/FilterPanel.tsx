import { FC } from 'react';
import * as style from './FilterPanel.styles';
import { TravelModeButton } from './TravelModeButton';
import { TravelTimeButton } from './TravelTimeButton';
import { DayOfWeekButton } from './DayOfWeekButton';
import { TimeOfDayButton } from './TimeOfDayButton';

type FilterPanelProps = {
  onSave: () => void;
};

const FilterPanel: FC<FilterPanelProps> = ({ onSave }: FilterPanelProps) => {
  return (
    <div className={style.panel}>
      <TravelModeButton />
      <TravelTimeButton />
      <DayOfWeekButton />
      <TimeOfDayButton />
      <button onClick={() => onSave()}>Save</button>
    </div>
  );
};

export { FilterPanel };
