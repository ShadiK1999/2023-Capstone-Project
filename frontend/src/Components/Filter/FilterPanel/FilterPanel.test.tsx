// eslint-disable-next-line import/named
import { screen, render } from '@testing-library/react';
import { FilterPanel } from './FilterPanel';
import { describe, it, expect, vi, afterEach, beforeEach } from 'vitest';
import { HomeContext, Filter, HomeState, initialState } from '../../../Contexts/Home/state';

const mockedFilter: Filter = {
  type: 'duration',
  mode: 'WALKING',
  value: 15,
  daysOfWeek: ['SUN'],
  openTime: 0,
  closeTime: 23,
};

const mockedState: HomeState = {
  ...initialState,
  filter: mockedFilter,
};

describe('Filter', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date(2023, 8, 4));
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('Matches DOM Snapshot', () => {
    const tree = render(
      <HomeContext.Provider value={mockedState}>
        <FilterPanel onSave={() => void 0} />
      </HomeContext.Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('Renders the correct isochrone input', () => {
    render(
      <HomeContext.Provider value={mockedState}>
        <FilterPanel onSave={() => void 0} />
      </HomeContext.Provider>,
    );

    const sundayButton = screen.getByText('sun');
    const saturdayButton = screen.getByText('sat');

    const walkingModeButton = screen.getByText('walking');
    const bicyclingModeButton = screen.getByText('cycling');

    expect(sundayButton.getAttribute('data-selected')).toBe('true');
    expect(saturdayButton.getAttribute('data-selected')).toBe('false');

    expect(walkingModeButton.getAttribute('data-selected')).toBe('true');
    expect(bicyclingModeButton.getAttribute('data-selected')).toBe('false');

    // travel time
    expect(screen.getByText('15 min')).toBeInTheDocument();
    // open hour
    expect(screen.getByText('12am')).toBeInTheDocument();
    // close hour
    expect(screen.getByText('11pm')).toBeInTheDocument();
  });
});
