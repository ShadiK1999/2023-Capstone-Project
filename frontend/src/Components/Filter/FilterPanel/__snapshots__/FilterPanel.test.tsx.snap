// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="f57ofvg"
      >
        <div
          class="flg5ft8"
        >
          <p>
            Travel Mode
          </p>
          <div
            class="fk0ntar"
          >
            <div
              class="f1wwzehz"
            >
              <button
                class="f1m8dfem"
                data-selected="true"
              >
                walking
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                cycling
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                driving
              </button>
            </div>
          </div>
        </div>
        <div
          class="flg5ft8"
        >
          <p>
            Travel Time
          </p>
          <div
            class="fk0ntar"
          >
            <div
              class="f110hs72"
            >
              <div
                class="ffvwc24"
              >
                <img
                  alt="MINUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                />
              </div>
              <div
                class="f1sz7xol"
              >
                15 min
              </div>
              <div
                class="f1g8elpx"
              >
                <img
                  alt="PLUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="flg5ft8"
        >
          <p>
            Opening Days
          </p>
          <div
            class="fk0ntar"
          >
            <div
              class="f1wwzehz"
            >
              <button
                class="f1m8dfem"
                data-selected="true"
              >
                sun
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                mon
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                tue
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                wed
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                thu
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                fri
              </button>
              <button
                class="fotahl3"
                data-selected="false"
              >
                sat
              </button>
            </div>
          </div>
        </div>
        <div
          class="f9vbxna"
        >
          <div
            class="flg5ft8"
          >
            <p>
              Open from...
            </p>
            <div
              class="fk0ntar"
            >
              <div
                class="f110hs72"
              >
                <div
                  class="ffvwc24"
                >
                  <img
                    alt="MINUS"
                    src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                  />
                </div>
                <div
                  class="f1sz7xol"
                >
                  12am
                </div>
                <div
                  class="f1g8elpx"
                >
                  <img
                    alt="PLUS"
                    src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="flg5ft8"
          >
            <p>
              Until...
            </p>
            <div
              class="fk0ntar"
            >
              <div
                class="f110hs72"
              >
                <div
                  class="ffvwc24"
                >
                  <img
                    alt="MINUS"
                    src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                  />
                </div>
                <div
                  class="f1sz7xol"
                >
                  11pm
                </div>
                <div
                  class="f1g8elpx"
                >
                  <img
                    alt="PLUS"
                    src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <button>
          Save
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="f57ofvg"
    >
      <div
        class="flg5ft8"
      >
        <p>
          Travel Mode
        </p>
        <div
          class="fk0ntar"
        >
          <div
            class="f1wwzehz"
          >
            <button
              class="f1m8dfem"
              data-selected="true"
            >
              walking
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              cycling
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              driving
            </button>
          </div>
        </div>
      </div>
      <div
        class="flg5ft8"
      >
        <p>
          Travel Time
        </p>
        <div
          class="fk0ntar"
        >
          <div
            class="f110hs72"
          >
            <div
              class="ffvwc24"
            >
              <img
                alt="MINUS"
                src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
              />
            </div>
            <div
              class="f1sz7xol"
            >
              15 min
            </div>
            <div
              class="f1g8elpx"
            >
              <img
                alt="PLUS"
                src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flg5ft8"
      >
        <p>
          Opening Days
        </p>
        <div
          class="fk0ntar"
        >
          <div
            class="f1wwzehz"
          >
            <button
              class="f1m8dfem"
              data-selected="true"
            >
              sun
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              mon
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              tue
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              wed
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              thu
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              fri
            </button>
            <button
              class="fotahl3"
              data-selected="false"
            >
              sat
            </button>
          </div>
        </div>
      </div>
      <div
        class="f9vbxna"
      >
        <div
          class="flg5ft8"
        >
          <p>
            Open from...
          </p>
          <div
            class="fk0ntar"
          >
            <div
              class="f110hs72"
            >
              <div
                class="ffvwc24"
              >
                <img
                  alt="MINUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                />
              </div>
              <div
                class="f1sz7xol"
              >
                12am
              </div>
              <div
                class="f1g8elpx"
              >
                <img
                  alt="PLUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="flg5ft8"
        >
          <p>
            Until...
          </p>
          <div
            class="fk0ntar"
          >
            <div
              class="f110hs72"
            >
              <div
                class="ffvwc24"
              >
                <img
                  alt="MINUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Minus.svg"
                />
              </div>
              <div
                class="f1sz7xol"
              >
                11pm
              </div>
              <div
                class="f1g8elpx"
              >
                <img
                  alt="PLUS"
                  src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <button>
        Save
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
